finality-aleph/src/sync/ticker.rs

```diff
@@ -0,0 +1,143 @@
+use tokio::time::{sleep, Duration, Instant};
+
+/// This struct is used for determining when we should broadcast justification so that it does not happen too often.
+pub struct BroadcastTicker {
+    last_broadcast: Instant,
+    current_periodic_timeout: Duration,
+    max_timeout: Duration,
+    min_timeout: Duration,
+}
+
+impl BroadcastTicker {
+    pub fn new(max_timeout: Duration, min_timeout: Duration) -> Self {
+        Self {
+            last_broadcast: Instant::now(),
+            current_periodic_timeout: max_timeout,
+            max_timeout,
+            min_timeout,
+        }
+    }
+
+    /// Returns whether we should broadcast right now if we just imported a justification.
+    /// If min_timeout elapsed since last broadcast, returns true, sets last broadcast to now and will
+    /// return true again if called after `self.min_timeout`.
+    /// If not, returns false and sets periodic broadcast timeout to `self.min_timeout`.
+    /// This is to prevent from sending every justification when importing a batch of them. This way,
+    /// when receiving batch of justifications we will broadcast the first justification and the highest known
+    /// after `self.min_timeout` using periodic broadcast.
+    pub fn try_broadcast(&mut self) -> bool {
+        let now = Instant::now();
+        if now.saturating_duration_since(self.last_broadcast) >= self.min_timeout {
+            self.last_broadcast = now;
```

Comment:
Huh, bug – you should set `timeout` to `max_timeout` here. Maybe even better, have a private method `tick` that sets the last tick and timeout.

Reply:
Oh, also make a regression test for that maybe – in this version as far as I can tell after a unsuccessful `try` followed by a successful one the next `wait` only takes `min_timeout` to fire.

Reply:
I thought it was supposed to work like that :P but yeah, there is no reason to broadcast soon after this is successful, so we can set here to `max_timeout`.